File: Leetcode-Solutions/LeetCode/Solutions-Python/daily/0611DailyTemperatures.py
# -*- coding: utf-8 -*-
# @Time : 2020/6/11 19:32
# @Author : Kevin
# @IDE: PyCharm
# @ModuleName : 0611DailyTemperatures
# @Description: 每日温度

class Solution:
    def dailyTemperatures_s1(self, T: List[int]) -> List[int]:
        n = len(T)
        ans, nxt, big = [0] * n, dict(), 10 ** 9
        for i in range(n - 1, -1, -1):
            warmer_index = min(nxt.get(t, big) for t in range(T[i] + 1, 102))
            if warmer_index != big:
                ans[i] = warmer_index - i
            nxt[T[i]] = i
        return ans

    # 单调栈
    def dailyTemperatures_s1(self, T: List[int]) -> List[int]:
        length = len(T)
        ans = [0] * length
        stack = []
        for i in range(length):
            temperature = T[i]
            while stack and temperature > T[stack[-1]]:
                prev_index = stack.pop()
                ans[prev_index] = i - prev_index
            stack.append(i)
        return ans

File: Leetcode-Solutions/LeetCode/Solutions-Golang/daily/main/0611dailyTemperatures.go
package main

import (
	"math"
)

func dailyTemperatures_s1(T []int) []int {
	length := len(T)
	ans := make([]int, length)
	next := make([]int, 101)
	for i := 0; i < 101; i++ {
		next[i] = math.MaxInt32
	}
	for i := length - 1; i >= 0; i-- {
		warmerIndex := math.MaxInt32
		for t := T[i] + 1; t <= 100; t++ {
			if next[t] < warmerIndex {
				warmerIndex = next[t]
			}
		}
		if warmerIndex < math.MaxInt32 {
			ans[i] = warmerIndex - i
		}
		next[T[i]] = i
	}
	return ans
}

func dailyTemperatures_s2(T []int) []int {
	length := len(T)
	ans := make([]int, length)
	stack := []int{}
	for i := 0; i < length; i++ {
		temperature := T[i]
		for len(stack) > 0 && temperature > T[stack[len(stack)-1]] { // 栈不为空 && 栈顶温度小于当前温度
			prevIndex := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			ans[prevIndex] = i - prevIndex
		}
		stack = append(stack, i)
	}
	return ans
}
